include/swift/AST/DiagnosticsSIL.def

```diff
@@ -320,6 +320,24 @@ ERROR(negative_fp_literal_overflow_unsigned, none,
   "negative literal '%0' cannot be converted to %select{|unsigned }2%1",
   (StringRef, Type, bool))
 
+// Overflow and underflow warnings for floating-point truncation
+WARNING(warning_float_trunc_overflow, none,
+  "overflow: '%0' becomes %select{|-}3inf during%select{| implicit}2 "
+  "conversion to %1", (StringRef, Type, bool, bool))
+
+WARNING(warning_float_trunc_underflow, none,
+  "precision loss due to tininess during%select{| implicit}2 "
+  "conversion of '%0' to %1", (StringRef, Type, bool, bool))
+
+WARNING(warning_float_trunc_hex_inexact, none,
+  "'%0' cannot be precisely represented by %1"
+  "%select{| or by smaller floating-point types}2",
```

Comment:
@xwu Here is some explanation for some of the points you raise: 

> Is "or by smaller floating-point types" necessary to make the essential point here?

  When you call initializers explicitly e.g. "Float(some_literal)"  -- first the literal is implicitly   converted to "Double" and then to "Float". So if there is a warning during the implicit conversion to "Double", the warning would refer to  the type "Double" which could be a bit surprising to the user since he/she never wrote it.  Btw, this problem is more general and not specific to Floating point conversion. A better diagnostic that takes in account the context of the implicit conversion is a bit more involved, and is an on going work. For the time being I think it is better to have some qualifier (even though redundant in some cases) that indicates that the type used in diagnostics and the one written by the user may be different.  I have made a recent change to the PR in which the error message reads slightly differently: "literal loses precision during (_implicit_) conversion to Double" , where the word (_implicit_) is included for warnings generated during implicit conversions but otherwise omitted. 

> ... non-representable by an IEEE floating-point type but representable by a hypothetical non-standard smaller floating-point type that has more bits for the significand.

 The diagnostics added here is very specific to the builtin floating-point  types supported by Swift: Float, Double and Float80 (on x86 architectures). They follow IEEE 754 standard. The diagnostics won't be emitted for any other user-defined Floating point types. 

> Can this warning be silenced by some syntax (say, parentheses, or an explicit casting notation such as T(0x1p2)) ?

The warning will be produced if and only if a floating-point type (Float, Double or Float80) is assigned or initialized with a hex float that is bound to lose precision when represented in the type. The warning wouldn't be  raised if the literal is first initialized to a variable of proper type and later cast to a smaller type.  So one way to eliminate the warning is to bind the literal to a large enough floating-point type and later convert it explicitly to a smaller type  e.g. 

`let x: Float =  0x1.00000000001p23` // warning will be produced
`let x = Float(0x1.00000000001p23)` // warning will be produced

`let x: Double =0x1.0000000001p23
  let y = Float(x)`  // no warning

> Indeed, that is how it works for decimal float literal notation and, even if not the default behavior, should not be something that becomes effectively prohibited for hex floats with an unsilenceable warning.

  We considered adding this warning for hex floats since the reason for using hex float notation  is to be precise and avoid any imprecision that may happen with decimal notation. If a truncation is desired, it could be done explicitly as illustrated by the second example mentioned above.